### PR TITLE
ENH: Clearer error when rejecting a solver.

### PR DESCRIPTION
**Pull request by ssanderson**
Provide a specific error indicating why a given solver was rejected
rather than a generic "could not solve problem" error.

Before/After:

```
(cvxpy) [~/quantopian/cvxpy]@(master:7deee4d172)$ python repro.py
Traceback (most recent call last):
  File "repro.py", line 11, in <module>
    problem.solve(solver='SCS')
  File "/home/ssanderson/quantopian/cvxpy/cvxpy/problems/problem.py", line 207, in solve
    return self._solve(*args, **kwargs)
  File "/home/ssanderson/quantopian/cvxpy/cvxpy/problems/problem.py", line 319, in _solve
    solver.validate_solver(constraints)
  File "/home/ssanderson/quantopian/cvxpy/cvxpy/problems/solvers/solver.py", line 137, in validate_solver
    "The solver %s cannot solve the problem." % self.name()
cvxpy.error.SolverError: The solver SCS cannot solve the problem.

(cvxpy) [~/quantopian/cvxpy]@(master:7deee4d172)$ pb
Switched to branch 'give-a-reason-when-solver-is-rejected'

(cvxpy) [~/quantopian/cvxpy]@(give-a-reason-when-solver-is-rejected:31982de1b0)$ python repro.py
Traceback (most recent call last):
  File "repro.py", line 11, in <module>
    problem.solve(solver='SCS')
  File "/home/ssanderson/quantopian/cvxpy/cvxpy/problems/problem.py", line 207, in solve
    return self._solve(*args, **kwargs)
  File "/home/ssanderson/quantopian/cvxpy/cvxpy/problems/problem.py", line 319, in _solve
    solver.validate_solver(constraints)
  File "/home/ssanderson/quantopian/cvxpy/cvxpy/problems/solvers/solver.py", line 139, in validate_solver
    self._reject_problem("it cannot solve unconstrained problems")
  File "/home/ssanderson/quantopian/cvxpy/cvxpy/problems/solvers/solver.py", line 158, in _reject_problem
    raise SolverError(message)
cvxpy.error.SolverError: The solver SCS cannot solve the problem because it cannot solve unconstrained problems.
```

### CODE REVIEW

**Comment by ssanderson** [comment]
Ran into this trying to figure out why some of my simplest tests were crashing when explicitly run with SCS.  I originally thought I was hitting some strange corner case, but it turned out that SCS doesn't support solving unconstrained problems.  This change provides separate error messages for the various cases that were already being enumerated in `Solver.validate_solver`.

File: cvxpy/problems/solvers/solver.py [hunk 1]
```diff
@@ -126,16 +126,36 @@ def validate_solver(self, constraints):
...
-           (len(constraints) == 0 and self.name() in [s.SCS,
-                                                      s.GLPK]):
-            raise SolverError(
-                "The solver %s cannot solve the problem." % self.name()
-            )
+
+        if (constr_map[s.BOOL] or constr_map[s.INT]) and not self.MIP_CAPABLE:
+            self._reject_problem("it cannot solve mixed-integer problems")
+        elif constr_map[s.SDP] and not self.SDP_CAPABLE:
+            self._reject_problem("it cannot solve semi-definite problems")
```
Comment (SteveDiamond) [inline cvxpy/problems/solvers/solver.py]: semidefinite

Comment (ssanderson) [inline cvxpy/problems/solvers/solver.py]: updated

**Comment by ssanderson** [comment]
AppVeyor failure looks like it's pre-existing?

**Comment by SteveDiamond** [comment]
Yeah, Appveyor has been broken for a long time. It's a very difficult situation with Windows and Python 3.5. This all looks good.
